extensions: Track the upstream extensions using meson subprojects

3v1n0 opened the pull request:
Manage all the extensions as a meson subproject, so that we can package
them atomically and ensure that the whole ecosystem works when tested.

Reduce patches when possible, using build system overrides.

This improves the maintainability of the extensions a lot.

Also added:
 - A simple tool to track dependencies
 - A tool to compute the shell dependencies
 - Autopkgtests 

/cc @aleasto

3v1n0 committed:
build: Include all supported extensions as meson subprojects

Manage all the extensions as a meson subproject, so that we can package
them atomically and ensure that the whole ecosystem works when tested.

Reduce patches when possible, using build system overrides

3v1n0 committed:
debian/control: Update standards versions

No changes needed

3v1n0 committed:
debian/control: Remove deprecated rules requires root

3v1n0 committed:
debian: Update packaging to use subprojects sources

Now that we only rely on meson subprojects for the extensions code, we
should use it to generate the source and build the packages.

So make it just as easy as use `gbp buildpackage` to ensure that
everything is handled automatically.

When that is triggered, the sources are downloaded and we ensure that
they are respecting the expected version

3v1n0 committed:
debian: Automate computing the required gnome-shell versions

Find the minimum and maximum gnome-shell versions we support by
inspecting all the metadata files that we have installed, so that there
is no risk that we produce a package that is installable if any of the
extension is not supported by the current gnome-shell

Gbp-Pq: Full

3v1n0 committed:
debian: Add simple tool to check and track runtime dependencies

Extensions may require external dependencies that can be discovered only
at runtime, so add some tooling that help to figure out the extensions
dependencies before it's too late

Gbp-Pq: full

3v1n0 committed:
debian: Only rely on gjs to check dependencies

It's the tool that eventually loads everything, so let's avoid depending
on both gi-inspect-typelib and gjs

3v1n0 committed:
debian/tests: Add autopkgtest

The tests are for now only covering upstream tests and dependency checks
but starting from GNOME 50, we could actually implement proper gnome
integration tests using the upstream tooling

jbicha commented:
I'm curious why you didn't want to keep separate binary packages.

I believe apt normally requires transitional packages in cases like this.

3v1n0 commented:
> I'm curious why you didn't want to keep separate binary packages.

Well, I was kinda undecided about those... Yes, we could still have separate packages here and bind them to the same package binary version.

Fact is that I would like that:
 - You get them all or none (extensions can still be disabled in software)
 - We can easily install a single package and test them all

> I believe apt normally requires transitional packages in cases like this.

Well, per se it does not need them (I can just install the new one and apt replaces the old ones properly), but once this goes in we should seed the whole package instead of the single packages.

jbicha commented:
I believe that even if `ubuntu-desktop` depends on `gnome-shell-ubuntu-extensions`, apt won't automatically install the updated packages when upgrading if it requires removing other packages. Transitional packages fixes that problem.

3v1n0 committed:
ci: Add workflow that checks the package build

3v1n0 commented:
> I believe that even if `ubuntu-desktop` depends on `gnome-shell-ubuntu-extensions`, apt won't automatically install the updated packages when upgrading if it requires removing other packages. Transitional packages fixes that problem.

Ok, but ideally I would like to remove them from the archive, wouldn't be that enough?

However, we can do even those in case

jbicha commented:
You can remove transitional packages after they have been included in an LTS. If you make the change before Ubuntu 26.04 LTS is released, you can remove the transitional packages in May for the Ubuntu 26.10 development cycle.

jbicha commented:
You would need your Breaks/Replaces to be versioned. It would be easiest for version number handling to use the existing source packages to build the transitional packages.

3v1n0 commented:
> You would need your Breaks/Replaces to be versioned. It would be easiest for version number handling to use the existing source packages to build the transitional packages.

Yeah, I was waiting to do that once this can go to main though, so once it's the time to do the transition from the other packages to this one, we can do the transitional ones and adjust this.

aleasto commented:
This could also:
```
Provides: gnome-shell-extension-appindicator,
          gnome-shell-extension-desktop-icons-ng,
          gnome-shell-extension-ubuntu-tiling-assistant,
          gnome-shell-extension-ubuntu-dock,
```

3v1n0 commented:
> Provides: gnome-shell-extension-appindicator,
>           gnome-shell-extension-desktop-icons-ng,
>           gnome-shell-extension-ubuntu-tiling-assistant,
>           gnome-shell-extension-ubuntu-dock,

Yep, I forgot about but I had it in a local stash, thanks!

3v1n0 commented:
Ah, also note that at the moment I did not bother changing the extensions code upstream, but ideally I'll move d2d and maybe tiling extension to meson too so that the subprojects structure here is even simpler.

3v1n0 committed:
debian/changelog: Update

aleasto commented:
> > Provides: gnome-shell-extension-appindicator,
> > gnome-shell-extension-desktop-icons-ng,
> > gnome-shell-extension-ubuntu-tiling-assistant,
> > gnome-shell-extension-ubuntu-dock,
> 
> Yep, I forgot about but I had it in a local stash, thanks!

Is it still in a local stash? :grin:

Actually if you make the Provides `${binary:version}` versioned  then I suppose these become effectively the transitional packages

jbicha commented:
> Actually if you make the Provides ${binary:version} versioned then I suppose these become effectively the transitional packages

I wish that worked but I believe we need actual transitional packages.

I don't think versioned Provides makes sense because the version numbers won't match upstream's version numbers.

3v1n0 committed:
debian/README.source: Explain the versioning schema

3v1n0 committed:
Update changelog

3v1n0 commented:
> > > Provides: gnome-shell-extension-appindicator,
> > > gnome-shell-extension-desktop-icons-ng,
> > > gnome-shell-extension-ubuntu-tiling-assistant,
> > > gnome-shell-extension-ubuntu-dock,
> > 
> > 
> > Yep, I forgot about but I had it in a local stash, thanks!
> 
> Is it still in a local stash? 😁

Sorry, forgot again :) (fixed now)

However, if there are no blockers, I'd like to get this to universe for now and we can then later handle the potential versioning checks once the current gnome-shell can depend on this.

aleasto commented:
The latest guidelines for versioning of Ubuntu-native packages suggest an `ubuntu` suffix: https://documentation.ubuntu.com/project/how-ubuntu-is-made/concepts/version-strings/#version-native-packages

aleasto commented on debian/control:
```diff
@@ -4,22 +4,62 @@ Priority: optional
+Provides: gnome-shell-extension-appindicator (= ${binary:Version}),
+          gnome-shell-extension-desktop-icons-ng (= ${binary:Version}),
+          gnome-shell-extension-ubuntu-tiling-assistant (= ${binary:Version}),
+          gnome-shell-extension-ubuntu-dock (= ${binary:Version}),
```
@jbicha was against using versioned-Provides because these do have version numbers upstream (and in Debian, perhaps more importantly) that do not match the versioning scheme used here.

3v1n0 replied:
Yeah I was undecided too, let's go to without.

3v1n0 replied:
done (and changed versioning schema too)

3v1n0 commented:
> The latest guidelines for versioning of Ubuntu-native packages suggest an `ubuntu` suffix: https://documentation.ubuntu.com/project/how-ubuntu-is-made/concepts/version-strings/#version-native-packages

This is relevant for packages that will also be in debian, while I don't see it happening for this... 

However if we really want we can go with:

`49.26.04ubuntu0`

EDIT: This is done now

aleasto commented:
> > The latest guidelines for versioning of Ubuntu-native packages suggest an `ubuntu` suffix: https://documentation.ubuntu.com/project/how-ubuntu-is-made/concepts/version-strings/#version-native-packages
> 
> This is relevant for packages that will also be in debian, while I don't see it happening for this...
> 
> However if we really want we can go with:
> 
> `49.26.04ubuntu0`
> 
> EDIT: This is done now

For a package that does not exist in debian, the guidelines state to use `49.26.04.0ubuntu` instead

3v1n0 commented:
> > > The latest guidelines for versioning of Ubuntu-native packages suggest an `ubuntu` suffix: https://documentation.ubuntu.com/project/how-ubuntu-is-made/concepts/version-strings/#version-native-packages
> > 
> > 
> > This is relevant for packages that will also be in debian, while I don't see it happening for this...
> > However if we really want we can go with:
> > `49.26.04ubuntu0`
> > EDIT: This is done now
> 
> For a package that does not exist in debian, the guidelines state to use `49.26.04.0ubuntu` instead

Changed again (I think I had done it initially, but I liked it less), however... This is really something that won't affect this project ever (while we do it with a real rationale in in yaru).

aleasto commented on debian/README.source:
```diff
@@ -0,0 +1,49 @@
+  |    +------------ Ubuntu Release version we target to (e.g. 25.10)
+  +----------------- GNOME Shell release major version (e.g. 49)
+
+So for example: 49.25.10ubuntu0
```
nit: 
```suggestion
So for example: 49.25.10.0ubuntu
```

3v1n0 replied:
Ouch

3v1n0 committed:
tools: Add validate-wrap-revisions

We want to make sure that wrap revisions are valid git tags or revisions
rather than branches so that they cannot be changed without noticing it.

Tags technically can be changed and we may also require their SHA
instead, but we will also be able to track this at upload time, so
let's simplify things for now
